### DISCUSSION TITLE
Remove the TFX dependency from the Model Card Toolkit

### DISCUSSION THREAD

**Discussion by hanneshapke** [tensorflow/model-card-toolkit#228] [Ideas]
The TFX Addons SIG wants to move the TFX component to the [TFX Addons repo](https://github.com/tensorflow/tfx-addons). This work allows us to remove the entire TFX dependencies from this repository.

The new dependencies could look like
```
REQUIRED_PACKAGES = [
    'absl-py>=0.9,<1.1',
    'jinja2>=3.1,<3.2',
    'matplotlib>=3.2.0,<4',
    'jsonschema>=3.2.0,<4',
    'tensorflow-data-validation>=1.5.0,<1.11.0',
    'tensorflow-model-analysis>=1.5.0,<1.11.0',
    'tensorflow-metadata>=1.5.0,<1.11.0',
    'ml-metadata>=1.5.0,<1.11.0',
    'dataclasses;python_version<"3.7"',
]
```

Furthermore, we should investigate if python version requirement still persists.
```
'dataclasses;python_version<"3.7"',
```

**Comment by codesue** [comment]
Summarizing discussions that happened elsewhere:
- We'll start by removing the plain TFX dependency and upgrading other libraries to unblock using MCT with newer versions of TFX. This is being done in https://github.com/tensorflow/model-card-toolkit/pull/237.
- Later, we can look into whether it's feasible to move functionality specific to MLMD, e.g. in [tfx_util.py](https://github.com/tensorflow/model-card-toolkit/blob/master/model_card_toolkit/utils/tfx_util.py), to the [tfx-addons MLMD client library](https://github.com/tensorflow/tfx-addons/tree/main/tfx_addons/mlmd_client). If so, we could make TFX and MLMD utils available through a `[tfx]` extra, for example.

**Comment by adrinjalali** [comment]
That still leaves a few tensorflow dependencies:

```
    'tensorflow-data-validation>=1.5.0,<1.11.0',
    'tensorflow-model-analysis>=1.5.0,<1.11.0',
    'tensorflow-metadata>=1.5.0,<1.11.0',
```
which I think are rather large. I suppose this library is only useful for people who use TF, and I personally don't, and don't necessarily want my envs to be so large.

Although one can produce a model card for a scikit-learn based model using this lib, having TF deps makes it not a viable option; which is fine, this is a TF centered lib, and one needs to know that. It'd be nice to make it more explicit in the main pages of the docs.

**Reply by hanneshapke** [reply]
Hi @adrinjalali,
Thank you for your feedback. We just started maintaining the package, therefore the MCT group is in requirements collection mode. Therefore, your thoughts are highly appreciated.

As @codesue mentioned, we'll remove the dependencies in multiple steps starting with TFX in the coming 1.4.0 release. If we would remove TFMA, how would you generate the metrics? Would you provide your own custom metrics functions? TFMA has a ton of dependencies, but it also does a ton of things out of the box (slicing, calc metrics, etc) and nowadays it works with all frameworks (incl. PyTorch).

Thank you for sharing your thoughts on how you would generate the metrics.

**Reply by adrinjalali** [reply]
I'm glad your team is now working on this. It's got great potential 🥳 

I agree and understand that having a tool to generate metrics is quite handy, but I don't think they have to be hard dependencies. There are several ways to generate metrics and corresponding images, and we could think of having different sub-modules which handle different libraries, or to give the control to the user and have good examples on how to produce them, kinda similar to this work we're doing: https://github.com/skops-dev/skops/pull/142

From the user's perspective, I think it'd be ideal for them to use the tools they'd usually have based on the model they pass. For instance, use PyTorch related tools for a pytorch model, a TF set of tools for a TF model, and a scikit-learn/fairlearn kinda work for a scikit-learn based model (this can include XGBoost/lightbgm).

WDYT?

**Reply by hanneshapke** [reply]
Great feedback @adrinjalali 

One thing upfront: There isn't a Google team supporting this work. :) We are all community contributors at this point.
I think we need to normalize/standardize the metric results somewhere in the MCT. That can be before or after the metrics generation. 

If you wouldn't use TFMA with PyTorch which tool chain would you use? Let's see if we can find a common denominator between the interface of TFMA, tools for PyTorch, etc. to be more tool agnostic. Please allow us a bit more time to compare tools and think about the design. Any insights are highly welcomed!

**Reply by adrinjalali** [reply]
Sure. I'm not very fluent with PyTorch or TensorFlow, but if you need help on the sklearn side, happy to give a hand.

**Reply by codesue** [reply]
For what it's worth, MCT doesn't generate metrics (i.e. evaluating models and outputting metrics). It (optionally) reads and parses precomputed metrics and adds them to a model card's list of `PerformanceMetric`s. See [source code](https://github.com/tensorflow/model-card-toolkit/blob/master/model_card_toolkit/utils/tfx_util.py#L403-L469). You can also do this manually, as in the [standalone example](https://www.tensorflow.org/responsible_ai/model_card_toolkit/examples/Standalone_Model_Card_Toolkit_Demo#quantitative_analysis) or read metrics from, e.g., a regular Python dictionary as in this [notebook](https://gist.github.com/codesue/2097704f3b84029f6e550e7f1650f378). Pretty much anything standalone/manual is framework agnostic but most automations are specific to TensorFlow. I haven't seen documentation on how to write your own utils to automate annotating metrics for other frameworks, though.

I really like the idea of having different sub-modules which handle different libraries. Elsewhere, someone also brought up the idea of limiting the main model-card-toolkit module to the bare minimum needed to define and export a model card.

**Reply by adrinjalali** [reply]
I remember trying to remove these dependencies from the repo when it came out, the issue was they were using TF data structures to pass things around.

If we change that to internal containers to store metric data, then it'd be easy to decouple these dependencies. There can be a module which reads info from TF files, and returns an internal container. Would need quite a bit of refactoring though.

**Reply by codesue** [reply]
Do you remember where TF data structures where being used to pass things around? If it was custom protos, these use `protobuf` and don't require TensorFlow and friends. I took a closer look, and in my understanding, now, TF data structures are only used in `ModelCardToolkit` if a `source` or `mlmd_source` is provided. Annotation functions are still called when these are `None`, but they're effectively no-ops.

**Reply by codesue** [reply]
I did some testing, and I think the minimal requirements to support core functionality with little refactoring are:

- `absl-py>=0.9,<1.1`: primarily used for testing and building docs, used for logging in one instance -- depending on how TensorFlow docs are built, it might be possible to move this to an extra 
- `jinja2>=3.1,<3.2`: used for rendering the model card
- `jsonschema>=3.2.0,<4`: used for validating JSON schema
- `protobuf>=3.19.0,<3.20.0`: used for building, serializing, and parsing protos -- could be made optional if exporting a model card as a proto is optional

Here, core functionality means creating `ModelCard` and `ModelCardToolkit` objects, only _manually_ annotating graphics and performance metrics, and rendering and exporting the model card. In other words, only the functionality necessary for running the [Standalone Model Card Toolkit Demo](https://www.tensorflow.org/responsible_ai/model_card_toolkit/examples/Standalone_Model_Card_Toolkit_Demo).

I made the following changes in this [quick, proof-of-concept branch](https://github.com/tensorflow/model-card-toolkit/compare/master...codesue:model-card-toolkit:codesue/do-not-pr-minimize-deps?expand=1) to try this out:

- updated required deps in `setup.py` and introduced a `tensorflow` extra 
- removed `from model_card_toolkit.utils.source import *` from the main `__init__.py` and the `utils/__init__.py`
- used optional imports in `core.py`
- moved tests in `core_test.py` that needed TensorFlow dependencies into a new test file

I tried it out by installing the minimal deps, running the demo notebook, and testing relevant files with pytest using:

```
pytest model_card_toolkit \              
--ignore model_card_toolkit/core_tensorflow_test.py \
--ignore model_card_toolkit/documentation/examples/cats_vs_dogs_test.py \
--ignore model_card_toolkit/utils/graphics_test.py \
--ignore model_card_toolkit/utils/tfx_util_test.py \
--ignore model_card_toolkit/utils/testdata/testdata_utils_test.py
```

Then I verified that all tests passed in another virtual environment that had TensorFlow deps.

Our next release will be a major one, so we have more wiggle room with making API changes that'll make it easier to support other frameworks. It's be great to have help with design ideas, feedback and code reviews. We're in the process of updating the contributing guide, so hopefully it'll become easier for people to help with implementation. 😄

**Reply by adrinjalali** [reply]
That looks perfect!

**Comment by codesue** [comment]
> Furthermore, we should investigate if python version requirement still persists.
> 
> ```
> 'dataclasses;python_version<"3.7"',
> ```

Let's bump the min Python version in [setup.py](https://github.com/tensorflow/model-card-toolkit/blob/master/setup.py#L126) to `3.7` and drop this requirement for the next release.

**Comment by codesue** [comment]
The `tfx` dependency was removed in https://github.com/tensorflow/model-card-toolkit/pull/248 and there's an issue for making `tensorflow-xyz` dependencies optional.
